# Introduction

In this notebook, we will do a comprehensive analysis of the Android app market by comparing thousands of apps in the Google Play store.

# About the Dataset of Google Play Store Apps & Reviews

**Data Source:** <br>
App and review data was scraped from the Google Play Store by Lavanya Gupta in 2018. Original files listed [here](
https://www.kaggle.com/lava18/google-play-store-apps).

# Import Statements

In [6]:
import pandas as pd


# Notebook Presentation

In [7]:
# Show numeric output in decimal format e.g., 2.15
pd.options.display.float_format = '{:,.2f}'.format

# Read the Dataset

In [8]:
df_apps = pd.read_csv('/content/drive/MyDrive/apps.csv')

In [9]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# Data Cleaning

**Challenge**: How many rows and columns does `df_apps` have? What are the column names? Look at a random sample of 5 different rows with [.sample()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html).

In [10]:
print(df_apps.shape)
df_apps.head()

(10841, 12)


,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Genres,Last_Updated,Android_Ver
0,Ak Parti Yardım Toplama,SOCIAL,NaN,0,8.70,0,Paid,$13.99,Teen,Social,"July 28, 2017",4.1 and up
1,Ain Arabic Kids Alif Ba ta,FAMILY,NaN,0,33.00,0,Paid,$2.99,Everyone,Education,"April 15, 2016",3.0 and up
2,Popsicle Launcher for Android P 9.0 launcher,PERSONALIZATION,NaN,0,5.50,0,Paid,$1.49,Everyone,Personalization,"July 11, 2018",4.2 and up
3,Command & Conquer: Rivals,FAMILY,NaN,0,19.00,0,NaN,0,Everyone 10+,Strategy,"June 28, 2018",Varies with device
4,CX Network,BUSINESS,NaN,0,10.00,0,Free,0,Everyone,Business,"August 6, 2018",4.1 and up


In [11]:
df_apps.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size_MBs', 'Installs', 'Type',
       'Price', 'Content_Rating', 'Genres', 'Last_Updated', 'Android_Ver'],
      dtype='object')

In [12]:

df_apps.sample(5)

,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Genres,Last_Updated,Android_Ver
1005,HD Themes Toyota Cruiser 200,PERSONALIZATION,3.70,3,15.00,100,Free,0,Teen,Personalization,"October 1, 2016",2.3.3 and up
9314,ESPN,SPORTS,4.20,521140,14.00,"10,000,000",Free,0,Everyone 10+,Sports,"July 19, 2018",5.0 and up
9157,Calorie Counter by FatSecret,HEALTH_AND_FITNESS,4.40,229210,8.80,"10,000,000",Free,0,Everyone,Health & Fitness,"July 31, 2018",Varies with device
4846,MOMOLAND BBoom BBoom,LIBRARIES_AND_DEMO,4.60,2580,20.00,"100,000",Free,0,Everyone,Libraries & Demo,"March 8, 2018",2.3 and up
1156,AG test,TOOLS,NaN,3,18.00,100,Free,0,Everyone,Tools,"October 18, 2017",4.1 and up


**Outcomes**: Compared to the previous projects we are working with a fairly large DataFrame this time. df_apps.shape tells us we have 10841 rows and 12 columns. We can already see that there are some data issues that we need to fix. In the Ratings and Type columns there are NaN (Not a number values) and in the Price column we have dollar signs that will cause problems. The sample(n) method will give us n random rows. This is another handy way to inspect our DataFrame.

### Drop Unused Columns

**Challenge**: Remove the columns called `Last_Updated` and `Android_Version` from the DataFrame. We will not use these columns.

In [13]:
df_apps = df_apps.drop(columns=['Last_Updated', 'Android_Ver'])
df_apps.sample(5)

,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Genres
2630,Critical Care Paramedic Review,MEDICAL,4.40,17,1.80,"1,000",Paid,$9.99,Everyone 10+,Medical
5018,REPUVE,AUTO_AND_VEHICLES,3.90,356,9.00,"100,000",Free,0,Everyone,Auto & Vehicles
4762,"muzmatch: Muslim & Arab Singles, Marriage & Da...",DATING,3.80,5168,25.00,"100,000",Free,0,Mature 17+,Dating
10828,YouTube,VIDEO_PLAYERS,4.30,25655305,4.65,"1,000,000,000",Free,0,Teen,Video Players & Editors
6585,Name Art Photo Editor - Focus n Filters,ART_AND_DESIGN,4.40,8788,12.00,"1,000,000",Free,0,Everyone,Art & Design


**Outcome**: To remove the unwanted columns, we simply provide a list of the column names ['Last_Updated', ‘Android_Ver'] to the .drop() method. and thats how we can remove columns from the dataframe

### Find and Remove NaN values in Ratings

**Challenge**: How may rows have a NaN value (not-a-number) in the Ratings column? Create DataFrame called `df_apps_clean` that does not include these rows.

In [14]:
nan_rows = df_apps[df_apps.Rating.isna()]
print(nan_rows.shape)
nan_rows.head()

(1474, 10)


,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Genres
0,Ak Parti Yardım Toplama,SOCIAL,NaN,0,8.70,0,Paid,$13.99,Teen,Social
1,Ain Arabic Kids Alif Ba ta,FAMILY,NaN,0,33.00,0,Paid,$2.99,Everyone,Education
2,Popsicle Launcher for Android P 9.0 launcher,PERSONALIZATION,NaN,0,5.50,0,Paid,$1.49,Everyone,Personalization
3,Command & Conquer: Rivals,FAMILY,NaN,0,19.00,0,NaN,0,Everyone 10+,Strategy
4,CX Network,BUSINESS,NaN,0,10.00,0,Free,0,Everyone,Business


In [15]:
df_apps_clean = df_apps.dropna()
df_apps_clean.isna().sum()
print(df_apps_clean.shape)

(9367, 10)


**Outcome**: To find and remove the rows with the NaN values we can create a subset of the DataFrame based on where .isna() evaluates to True. We see that NaN values in ratings are associated with no reviews (and no installs). That makes sense.We can drop the NaN values with .dropna() This leaves us with 9,367 entries in our DataFrame

### Find and Remove Duplicates

**Challenge**: Are there any duplicates in data? Check for duplicates using the [.duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) function. How many entries can you find for the "Instagram" app? Use [.drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) to remove any duplicates from `df_apps_clean`.


In [16]:
duplicated_rows = df_apps_clean[df_apps_clean.duplicated()]
print(duplicated_rows.shape)
duplicated_rows.head()

(476, 10)


,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Genres
946,420 BZ Budeze Delivery,MEDICAL,5.00,2,11.00,100,Free,0,Mature 17+,Medical
1133,MouseMingle,DATING,2.70,3,3.90,100,Free,0,Mature 17+,Dating
1196,"Cardiac diagnosis (heart rate, arrhythmia)",MEDICAL,4.40,8,6.50,100,Paid,$12.99,Everyone,Medical
1231,Sway Medical,MEDICAL,5.00,3,22.00,100,Free,0,Everyone,Medical
1247,Chat Kids - Chat Room For Kids,DATING,4.70,6,4.90,100,Free,0,Mature 17+,Dating


In [17]:

df_apps_clean[df_apps_clean['App'] == 'Instagram']

,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Genres
10806,Instagram,SOCIAL,4.50,66577313,5.30,"1,000,000,000",Free,0,Teen,Social
10808,Instagram,SOCIAL,4.50,66577446,5.30,"1,000,000,000",Free,0,Teen,Social
10809,Instagram,SOCIAL,4.50,66577313,5.30,"1,000,000,000",Free,0,Teen,Social
10810,Instagram,SOCIAL,4.50,66509917,5.30,"1,000,000,000",Free,0,Teen,Social


**Outcome**: There are indeed duplicates in the data. We can show them using the .duplicated() method We can actually check for an individual app like ‘Instagram’ by looking up all the entries with that name in the App column. just like we have done it above

In [18]:
df_apps_clean = df_apps_clean.drop_duplicates()
print(df_apps_clean.duplicated().sum())
df_apps_clean[df_apps_clean['App'] == 'Instagram']

0


,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Genres
10806,Instagram,SOCIAL,4.50,66577313,5.30,"1,000,000,000",Free,0,Teen,Social
10808,Instagram,SOCIAL,4.50,66577446,5.30,"1,000,000,000",Free,0,Teen,Social
10810,Instagram,SOCIAL,4.50,66509917,5.30,"1,000,000,000",Free,0,Teen,Social


In [19]:
df_apps_clean = df_apps_clean.drop_duplicates(subset = ['App', 'Type', 'Price'])
print(df_apps_clean.duplicated().sum())
df_apps_clean[df_apps_clean['App'] == 'Instagram']

0


,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Genres
10806,Instagram,SOCIAL,4.50,66577313,5.30,"1,000,000,000",Free,0,Teen,Social


In [20]:
df_apps_clean.shape

(8199, 10)

**Outcome**: So how do we get rid of duplicates? Can we simply call.drop_duplicates()? Not really. If we do this without specifying how to identify duplicates, we see that 3 copies of Instagram are retained because they have a different number of reviews. We need to provide the column names that should be used in the comparison to identify duplicates

# Find Highest Rated Apps

**Challenge**: Identify which apps are the highest rated. What problem might you encounter if you rely exclusively on ratings alone to determine the quality of an app?

In [21]:
print(df_apps_clean.Rating.idxmax())

df_apps.loc[21]

21


,21
App,KBA-EZ Health Guide
Category,MEDICAL
Rating,5.00
Reviews,4
Size_MBs,25.00
Installs,1
Type,Free
Price,0
Content_Rating,Everyone
Genres,Medical


In [22]:
df_apps_clean.sort_values('Rating', ascending=False).head()

,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Genres
21,KBA-EZ Health Guide,MEDICAL,5.00,4,25.00,1,Free,0,Everyone,Medical
1751,Spine- The dating app,DATING,5.00,5,9.30,500,Free,0,Teen,Dating
1761,Selfie With Champion AJ Style,PHOTOGRAPHY,5.00,2,7.50,500,Free,0,Everyone,Photography
1763,Ag Valley Cooperative,BUSINESS,5.00,6,74.00,500,Free,0,Everyone,Business
1769,Yazdani Cd Center EllahAbad Official App,FAMILY,5.00,8,3.80,500,Free,0,Everyone,Entertainment


**Outcome**: Only apps with very few reviews (and a low number on installs) have perfect 5 star ratings (most likely by friends and family).

# Find 5 Largest Apps in terms of Size (MBs)

**Challenge**: What's the size in megabytes (MB) of the largest Android apps in the Google Play Store. Based on the data, do you think there could be limit in place or can developers make apps as large as they please?

In [23]:
df_apps_clean.sort_values('Size_MBs', ascending=False).head()

,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Genres
8719,Draft Simulator for FUT 18,SPORTS,4.60,162933,100.00,"5,000,000",Free,0,Everyone,Sports
8718,Mini Golf King - Multiplayer Game,GAME,4.50,531458,100.00,"5,000,000",Free,0,Everyone,Sports
10295,SimCity BuildIt,FAMILY,4.50,4218587,100.00,"50,000,000",Free,0,Everyone 10+,Simulation
9945,Ultimate Tennis,SPORTS,4.30,183004,100.00,"10,000,000",Free,0,Everyone,Sports
9944,Gangster Town: Vice District,FAMILY,4.30,65146,100.00,"10,000,000",Free,0,Mature 17+,Simulation


**Outcome**: Here we can clearly see that there seems to be an upper bound of 100 MB for the size of an app. A quick google search would also have revealed that this limit is imposed by the Google Play Store itself. It’s interesting to see that a number of apps actually hit that limit exactly.

# Find the 5 App with Most Reviews

**Challenge**: Which apps have the highest number of reviews? Are there any paid apps among the top 50?

In [24]:
df_apps_clean.sort_values('Reviews', ascending=False).head(50)

,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Genres
10805,Facebook,SOCIAL,4.10,78158306,5.30,"1,000,000,000",Free,0,Teen,Social
10785,WhatsApp Messenger,COMMUNICATION,4.40,69119316,3.50,"1,000,000,000",Free,0,Everyone,Communication
10806,Instagram,SOCIAL,4.50,66577313,5.30,"1,000,000,000",Free,0,Teen,Social
10784,Messenger – Text and Video Chat for Free,COMMUNICATION,4.00,56642847,3.50,"1,000,000,000",Free,0,Everyone,Communication
10650,Clash of Clans,GAME,4.60,44891723,98.00,"100,000,000",Free,0,Everyone 10+,Strategy
10744,Clean Master- Space Cleaner & Antivirus,TOOLS,4.70,42916526,3.40,"500,000,000",Free,0,Everyone,Tools
10835,Subway Surfers,GAME,4.50,27722264,76.00,"1,000,000,000",Free,0,Everyone 10+,Arcade
10828,YouTube,VIDEO_PLAYERS,4.30,25655305,4.65,"1,000,000,000",Free,0,Teen,Video Players & Editors
10746,"Security Master - Antivirus, VPN, AppLock, Boo...",TOOLS,4.70,24900999,3.40,"500,000,000",Free,0,Everyone,Tools
10584,Clash Royale,GAME,4.60,23133508,97.00,"100,000,000",Free,0,Everyone 10+,Strategy


**Outcomoe**:If you look at the number of reviews, you can find the most popular apps on the Android App Store. These include the usual suspects: Facebook, WhatsApp, Instagram etc. What’s also notable is that the list of the top 50 most reviewed apps does not include a single paid app

# Plotly Pie and Donut Charts - Visualise Categorical Data: Content Ratings

In [25]:
ratings = df_apps_clean.Content_Rating.value_counts()
print(ratings)

Content_Rating
Everyone           6621
Teen                912
Mature 17+          357
Everyone 10+        305
Adults only 18+       3
Unrated               1
Name: count, dtype: int64


In [26]:
import plotly.express as px

In [27]:
# fig = px.pie(labels= ratings.index, values = ratings.values)
fig = px.pie(labels=ratings.index, values=ratings.values, title="Content Rating", names=ratings.index,)

In [28]:
fig.update_traces(textposition='outside', textinfo='percent+label')
fig.show()

In [29]:
fig = px.pie(labels=ratings.index, values=ratings.values,title="Content Rating", names=ratings.index, hole=0.6,)
fig.update_traces(textposition='inside', textfont_size=15, textinfo='percent')

fig.show()

# Numeric Type Conversion: Examine the Number of Installs

**Challenge**: How many apps had over 1 billion (that's right - BILLION) installations? How many apps just had a single install?

Check the datatype of the Installs column.

Count the number of apps at each level of installations.

Convert the number of installations (the Installs column) to a numeric data type. Hint: this is a 2-step process. You'll have make sure you remove non-numeric characters first.

In [30]:
# Challenge: How many apps had over 1 billion (that's right - BILLION) installations? How many apps just had a single install?

billion_installs = df_apps_clean[df_apps_clean.Installs == '1,000,000,000'].value_counts()
one_installs = df_apps_clean[df_apps_clean.Installs == '1'].value_counts()
print(f"billion installs: \n{billion_installs}\n\nOne installs:\n {one_installs}\nTotal number of installs over billion are {len(billion_installs)} and total number of 1 installs are {len(one_installs)}")

billion installs: 
App                                       Category             Rating  Reviews   Size_MBs  Installs       Type  Price  Content_Rating  Genres                 
Facebook                                  SOCIAL               4.10    78158306  5.30      1,000,000,000  Free  0      Teen            Social                     1
Gmail                                     COMMUNICATION        4.30    4604324   3.50      1,000,000,000  Free  0      Everyone        Communication              1
Google                                    TOOLS                4.40    8033493   3.40      1,000,000,000  Free  0      Everyone        Tools                      1
Google Chrome: Fast & Secure              COMMUNICATION        4.30    9642995   3.50      1,000,000,000  Free  0      Everyone        Communication              1
Google Drive                              PRODUCTIVITY         4.40    2731171   4.00      1,000,000,000  Free  0      Everyone        Productivity               1
Go

In [31]:
# Check the datatype of the Installs column.

df_apps_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8199 entries, 21 to 10835
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             8199 non-null   object 
 1   Category        8199 non-null   object 
 2   Rating          8199 non-null   float64
 3   Reviews         8199 non-null   int64  
 4   Size_MBs        8199 non-null   float64
 5   Installs        8199 non-null   object 
 6   Type            8199 non-null   object 
 7   Price           8199 non-null   object 
 8   Content_Rating  8199 non-null   object 
 9   Genres          8199 non-null   object 
dtypes: float64(2), int64(1), object(7)
memory usage: 704.6+ KB


**Outcome**:  these show that we are dealing with a non-numeric data type. In this case, the type is "object".

In [32]:
# Count the number of apps at each level of installations.?

installation_milestones = df_apps_clean.Installs.value_counts()
print(installation_milestones)

Installs
1,000,000        1417
100,000          1096
10,000            988
10,000,000        933
1,000             698
5,000,000         607
500,000           504
50,000            457
5,000             425
100               303
50,000,000        202
500               199
100,000,000       189
10                 69
50                 56
500,000,000        24
1,000,000,000      20
5                   9
1                   3
Name: count, dtype: int64


In [33]:
# Convert the number of installations (the Installs column) to a numeric data type. Hint: this is a 2-step process. You'll have make sure you remove non-numeric characters first.

df_apps_clean.Installs = df_apps_clean.Installs.str.replace(',', '').astype(int)

In [34]:
df_apps_clean.sample(10)

,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Genres
9211,ABC Kids - Tracing & Phonics,FAMILY,4.50,36606,19.00,10000000,Free,0,Everyone,Educational;Education
2090,Hi-Smart AC,TOOLS,2.20,32,9.40,1000,Free,0,Everyone,Tools
8766,Kids Balloon Pop Game Free 🎈,FAMILY,4.10,38021,19.00,10000000,Free,0,Everyone,Casual;Music & Video
6096,Mod GTA V for MCPE,FAMILY,3.30,4288,10.00,500000,Free,0,Everyone,Entertainment
6480,Jewels Star: OZ adventure,GAME,4.50,21892,14.00,1000000,Free,0,Everyone,Puzzle
1407,LTC AS Legal,MEDICAL,4.00,6,1.30,100,Paid,$39.99,Everyone,Medical
2136,GKPB FP Online Church,LIFESTYLE,5.00,32,7.90,1000,Free,0,Everyone,Lifestyle
9444,Home Workout - No Equipment,HEALTH_AND_FITNESS,4.80,428156,15.00,10000000,Free,0,Everyone,Health & Fitness
6381,Swift for Facebook Lite,SOCIAL,4.30,9562,5.80,500000,Free,0,Teen,Social
1911,Fan App for Portsmouth FC,SPORTS,4.50,53,12.00,1000,Free,0,Teen,Sports


**Outcome**: We can remove the comma (,) character - or any character for that matter - from a DataFrame using the string’s .replace() method. Here we’re saying: “replace the , with an empty string”. This completely removes all the commas in the Installs column. We can then convert our data to a number using astype(int) method

In [35]:
installation_milestones = df_apps_clean.Installs.value_counts()
print(installation_milestones)

Installs
1000000       1417
100000        1096
10000          988
10000000       933
1000           698
5000000        607
500000         504
50000          457
5000           425
100            303
50000000       202
500            199
100000000      189
10              69
50              56
500000000       24
1000000000      20
5                9
1                3
Name: count, dtype: int64


# Find the Most Expensive Apps, Filter out the Junk, and Calculate a (ballpark) Sales Revenue Estimate

Let's examine the Price column more closely.

**Challenge**: Convert the price column to numeric data. Then investigate the top 20 most expensive apps in the dataset.

Remove all apps that cost more than $250 from the `df_apps_clean` DataFrame.

Add a column called 'Revenue_Estimate' to the DataFrame. This column should hold the price of the app times the number of installs. What are the top 10 highest grossing paid apps according to this estimate? Out of the top 10 highest grossing paid apps, how many are games?


In [36]:
df_apps_clean.sample(7)

,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Genres
1588,Best CG Backgrounds,FAMILY,3.60,8,3.70,500,Free,0,Everyone,Entertainment
8907,ZOOM Cloud Meetings,BUSINESS,4.40,31614,37.00,10000000,Free,0,Everyone,Business
4514,Freecell CY,GAME,4.00,387,1.10,50000,Free,0,Everyone,Card
3300,Birmingham City FC,SPORTS,4.50,191,28.00,10000,Free,0,Everyone,Sports
4057,R+F PULSE,BUSINESS,3.40,58,3.00,10000,Free,0,Everyone,Business
4430,AC Remote for Sharp - NOW FREE,TOOLS,3.70,332,26.00,50000,Free,0,Everyone,Tools
7877,SW-100.tch by Callstel,COMMUNICATION,4.00,1926,0.70,1000000,Free,0,Everyone,Communication


In [37]:

# Convert the price column to numeric data. Then investigate the top 20 most expensive apps in the dataset.

df_apps_clean.Price = df_apps_clean.Price.str.replace('$', '').str.replace(',', '').astype(float)

In [38]:
df_apps_clean.Price.info()

<class 'pandas.core.series.Series'>
Index: 8199 entries, 21 to 10835
Series name: Price
Non-Null Count  Dtype  
--------------  -----  
8199 non-null   float64
dtypes: float64(1)
memory usage: 128.1 KB


In [39]:
df_apps_clean.sort_values('Price', ascending=False).head(20)

,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Genres
3946,I'm Rich - Trump Edition,LIFESTYLE,3.60,275,7.30,10000,Paid,400.00,Everyone,Lifestyle
2394,I am Rich!,FINANCE,3.80,93,22.00,1000,Paid,399.99,Everyone,Finance
2461,I AM RICH PRO PLUS,FINANCE,4.00,36,41.00,1000,Paid,399.99,Everyone,Finance
1331,most expensive app (H),FAMILY,4.30,6,1.50,100,Paid,399.99,Everyone,Entertainment
3145,I am rich(premium),FINANCE,3.50,472,0.94,5000,Paid,399.99,Everyone,Finance
2775,I Am Rich Pro,FAMILY,4.40,201,2.70,5000,Paid,399.99,Everyone,Entertainment
5765,I am rich,LIFESTYLE,3.80,3547,1.80,100000,Paid,399.99,Everyone,Lifestyle
3221,I am Rich Plus,FAMILY,4.00,856,8.70,10000,Paid,399.99,Everyone,Entertainment
3554,💎 I'm rich,LIFESTYLE,3.80,718,26.00,10000,Paid,399.99,Everyone,Lifestyle
3114,I am Rich,FINANCE,4.30,180,3.80,5000,Paid,399.99,Everyone,Finance


**Outcome**: You also see that is of type object. The reason is the dollar $ signs that we’ve spotted before. To convert the price column to numeric data we use the .replace() method once again, but this time we filter out the dollar sign. What’s going on here? There are 15 I am Rich Apps in the Google Play Store apparently. They all cost 300 dollar or more, which is the main point of the app. The story goes that in 2008, Armin Heinrich released the very first I am Rich app in the iOS App Store for 999.90 dollar. The app does absolutely nothing. It just displays the picture of a gemstone and can be used to prove to your friends how rich you are. Armin actually made a total of 7 sales before the app was hastily removed by Apple. Nonetheless, it inspired a bunch of copycats on the Android App Store, but if you search today, you’ll find all of these apps have disappeared as well. The high installation numbers are likely gamed by making the app was available for free at some point to get reviews and appear more legitimate.**bold text**

### The most expensive apps sub $250

In [40]:
# Remove all apps that cost more than $250 from the df_apps_clean DataFrame.

df_apps_clean = df_apps_clean.drop(df_apps_clean[df_apps_clean.Price > 250].index)


In [41]:
df_apps_clean.sort_values('Price', ascending=False).head(5)

,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Genres
2281,Vargo Anesthesia Mega App,MEDICAL,4.60,92,32.00,1000,Paid,79.99,Everyone,Medical
1407,LTC AS Legal,MEDICAL,4.00,6,1.30,100,Paid,39.99,Everyone,Medical
2629,I am Rich Person,LIFESTYLE,4.20,134,1.80,1000,Paid,37.99,Everyone,Lifestyle
2481,A Manual of Acupuncture,MEDICAL,3.50,214,68.00,1000,Paid,33.99,Everyone,Medical
504,AP Art History Flashcards,FAMILY,5.00,1,96.00,10,Paid,29.99,Mature 17+,Education


### Highest Grossing Paid Apps (ballpark estimate)

In [42]:
Revenue_Estimate_col = df_apps_clean.Installs * df_apps_clean.Price
df_apps_clean.insert(9, 'Revenue_Estimate', Revenue_Estimate_col)
df_apps_clean.head()

,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Revenue_Estimate,Genres
21,KBA-EZ Health Guide,MEDICAL,5.00,4,25.00,1,Free,0.00,Everyone,0.00,Medical
28,Ra Ga Ba,GAME,5.00,2,20.00,1,Paid,1.49,Everyone,1.49,Arcade
47,Mu.F.O.,GAME,5.00,2,16.00,1,Paid,0.99,Everyone,0.99,Arcade
82,Brick Breaker BR,GAME,5.00,7,19.00,5,Free,0.00,Everyone,0.00,Arcade
99,Anatomy & Physiology Vocabulary Exam Review App,MEDICAL,5.00,1,4.60,5,Free,0.00,Everyone,0.00,Medical


**Outcome**: We can work out the highest grossing paid apps now. All we need to do is multiply the values in the price and the installs column to get the number

In [43]:
df_apps_clean.sample(7)

,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Revenue_Estimate,Genres
2699,Advanced EX for MITSUBISHI,TOOLS,2.90,33,0.14,1000,Paid,4.99,Everyone,"4,990.00",Tools
9714,Antivirus & Mobile Security,TOOLS,4.60,351267,5.00,10000000,Free,0.00,Everyone,0.00,Tools
1838,Movie Downloader Torrent : Az Torrent,VIDEO_PLAYERS,4.70,70,14.00,1000,Free,0.00,Everyone,0.00,Video Players & Editors
9604,MadLipz,FAMILY,4.60,171017,13.00,10000000,Free,0.00,Teen,0.00,Entertainment
3916,CG Shikshak,FAMILY,4.40,211,6.80,10000,Free,0.00,Everyone,0.00,Education
4452,C++ Tutorials,EDUCATION,4.10,358,1.90,50000,Free,0.00,Everyone,0.00,Education
1115,Mad Dash Fo' Cash,GAME,5.00,14,16.00,100,Free,0.00,Everyone,0.00,Arcade


In [44]:
# What are the top 10 highest-grossing paid apps according to this estimate? Out of the top 10, how many are games?
df_apps_clean.sort_values('Revenue_Estimate', ascending=False).head(10)

,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Revenue_Estimate,Genres
9220,Minecraft,FAMILY,4.50,2376564,19.00,10000000,Paid,6.99,Everyone 10+,"69,900,000.00",Arcade;Action & Adventure
8825,Hitman Sniper,GAME,4.60,408292,29.00,10000000,Paid,0.99,Mature 17+,"9,900,000.00",Action
7151,Grand Theft Auto: San Andreas,GAME,4.40,348962,26.00,1000000,Paid,6.99,Mature 17+,"6,990,000.00",Action
7477,Facetune - For Free,PHOTOGRAPHY,4.40,49553,48.00,1000000,Paid,5.99,Everyone,"5,990,000.00",Photography
7977,Sleep as Android Unlock,LIFESTYLE,4.50,23966,0.85,1000000,Paid,5.99,Everyone,"5,990,000.00",Lifestyle
6594,DraStic DS Emulator,GAME,4.60,87766,12.00,1000000,Paid,4.99,Everyone,"4,990,000.00",Action
6082,Weather Live,WEATHER,4.50,76593,4.75,500000,Paid,5.99,Everyone,"2,995,000.00",Weather
6746,Card Wars - Adventure Time,FAMILY,4.30,129603,23.00,1000000,Paid,2.99,Everyone 10+,"2,990,000.00",Card;Action & Adventure
7633,Five Nights at Freddy's,GAME,4.60,100805,50.00,1000000,Paid,2.99,Teen,"2,990,000.00",Action
7954,Bloons TD 5,FAMILY,4.60,190086,94.00,1000000,Paid,2.99,Everyone,"2,990,000.00",Strategy


**Outcome**:The top spot of the highest-grossing paid app goes to … Minecraft at close to $70 million. It’s quite interesting that Minecraft (along with Bloons and Card Wars) is actually listed in the Family category rather than in the Game category. If we include these titles, we see that 7 out the top 10 highest-grossing apps are games.

# Plotly Bar Charts & Scatter Plots: Analysing App Categories

In [45]:
df_apps_clean.Category.nunique() #We can find the number of different categories like so

33

In [46]:
top10_category = df_apps_clean.Category.value_counts()[:10] #To calculate the number of apps per category we can use our old friend .value_counts()

In [47]:
top10_category


,count
Category,
FAMILY,1606
GAME,910
TOOLS,719
PRODUCTIVITY,301
PERSONALIZATION,298
LIFESTYLE,297
FINANCE,296
MEDICAL,292
PHOTOGRAPHY,263


### Vertical Bar Chart - Highest Competition (Number of Apps)

In [48]:
bar = px.bar(x = top10_category.index, # index = category name
             y = top10_category.values)

bar.show()

**Outcome** Based on the number of apps, the Family and Game categories are the most competitive. Releasing yet another app into these categories will make it hard to get noticed.

### Horizontal Bar Chart - Most Popular Categories (Highest Downloads)

In [49]:
category_installs = df_apps_clean.groupby('Category').agg({'Installs': pd.Series.sum})
category_installs.sort_values('Installs', ascending=True, inplace=True)

In [50]:
h_bar = px.bar(x = category_installs.Installs,
               y = category_installs.index,
               orientation='h',
               title='Category Popularity')

h_bar.update_layout(xaxis_title='Number of Downloads', yaxis_title='Category')
h_bar.show()

### Category Concentration - Downloads vs. Competition

**Challenge**:
* First, create a DataFrame that has the number of apps in one column and the number of installs in another:

<img src=https://imgur.com/uQRSlXi.png width="350">

* Then use the [plotly express examples from the documentation](https://plotly.com/python/line-and-scatter/) alongside the [.scatter() API reference](https://plotly.com/python-api-reference/generated/plotly.express.scatter.html)to create scatter plot that looks like this.

<img src=https://imgur.com/cHsqh6a.png>

*Hint*: Use the size, hover_name and color parameters in .scatter(). To scale the yaxis, call .update_layout() and specify that the yaxis should be on a log-scale like so: yaxis=dict(type='log')

In [51]:
# First, create a DataFrame that has the number of apps in one column and the number of installs in another:
Category = df_apps_clean.Category.value_counts()

df_category = Category.to_frame(name='App')
df_category

,App
Category,
FAMILY,1606
GAME,910
TOOLS,719
PRODUCTIVITY,301
PERSONALIZATION,298
LIFESTYLE,297
FINANCE,296
MEDICAL,292
PHOTOGRAPHY,263


In [52]:
Installs_col = category_installs.Installs
df_category.insert(1, 'Installs', Installs_col)


In [53]:
print(df_category.shape)
df_category.sort_values('Installs', ascending=False)

(33, 2)


,App,Installs
Category,,
GAME,910,13858762717
COMMUNICATION,257,11039241530
TOOLS,719,8099724500
PRODUCTIVITY,301,5788070180
SOCIAL,203,5487841475
PHOTOGRAPHY,263,4649143130
FAMILY,1606,4437554490
VIDEO_PLAYERS,148,3916897200
TRAVEL_AND_LOCAL,187,2894859300


In [54]:
import plotly.express as px


In [55]:
df = px.data.iris()
scatter = px.scatter(df_category, # data
                    x='App', # column name
                    y='Installs',
                    title='Category Concentration',
                    size='App',
                    hover_name=df_category.index,
                    color='Installs')

scatter.show()

In [56]:
scatter.update_layout(xaxis_title="Number of Apps (Lower=More Concentrated)",
                      yaxis_title="Installs",
                      yaxis=dict(type='log'))

scatter.show()

**Outcome**: What we see is that the categories like Family, Tools, and Game have many different apps sharing a high number of downloads. But for the categories like video players and entertainment, all the downloads are concentrated in very few apps.

# Extracting Nested Data from a Column

**Challenge**: How many different types of genres are there? Can an app belong to more than one genre? Check what happens when you use .value_counts() on a column with nested values? See if you can work around this problem by using the .split() function and the DataFrame's [.stack() method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.stack.html).


In [57]:
print(f'There are total {df_apps_clean.Genres.nunique()} Genres in a dataframe')



There are total 114 Genres in a dataframe


In [58]:
df_apps_clean.sample(10)

,App,Category,Rating,Reviews,Size_MBs,Installs,Type,Price,Content_Rating,Revenue_Estimate,Genres
10639,Talking Tom Cat 2,FAMILY,4.30,3213548,55.00,100000000,Free,0.00,Everyone 10+,0.00,Entertainment
5892,EliteSingles – Dating for Single Professionals,DATING,2.50,5377,19.00,500000,Free,0.00,Mature 17+,0.00,Dating
1026,BL IP-Camera,TOOLS,4.30,15,1.20,100,Paid,3.99,Everyone,399.00,Tools
8164,"IndiaMART: Search Products, Buy, Sell & Trade",BUSINESS,4.50,207372,11.00,5000000,Free,0.00,Everyone,0.00,Business
9690,Kids Animals Jigsaw Puzzles 😄,FAMILY,4.10,51895,22.00,10000000,Free,0.00,Everyone,0.00,Educational;Brain Games
4650,"WeatherClear - Ad-free Weather, Minute forecast",WEATHER,4.50,3252,3.80,50000,Free,0.00,Everyone,0.00,Weather
2079,m>notes notepad,PRODUCTIVITY,4.30,184,4.00,1000,Paid,2.99,Everyone,"2,990.00",Productivity
1768,Endurance Lifestyle,EVENTS,4.60,7,3.80,500,Free,0.00,Teen,0.00,Events
9000,Moto Rider,GAME,4.20,115176,24.00,10000000,Free,0.00,Teen,0.00,Racing
9463,Guns - Shot Sounds,FAMILY,3.80,29544,10.00,10000000,Free,0.00,Everyone,0.00,Entertainment


**Outcome**: Yes An App can belong to more than 1 Genre as shown in the above table

In [59]:
df_apps_clean.Genres.value_counts()

,count
Genres,
Tools,718
Entertainment,467
Education,429
Productivity,301
Personalization,298
...,...
Casual;Music & Video,1
Tools;Education,1
Role Playing;Brain Games,1


In [60]:
# Split the strings on the semi-colon and then .stack them.
stack = df_apps_clean.Genres.str.split(';', expand=True).stack()
print(f'We now have a single column with shape: {stack.shape}')
num_genres = stack.value_counts()
print(f'Number of genres: {len(num_genres)}')

We now have a single column with shape: (8564,)
Number of genres: 53


**Outcome**: I somehow need to separate the genre names to get a clear picture. This is where the string’s .split() method comes in handy. After we’ve separated our genre names based on the semi-colon, we can add them all into a single column with .stack() and then use .value_counts().

# Colour Scales in Plotly Charts - Competition in Genres

**Challenge**: Can you create this chart with the Series containing the genre data?

<img src=https://imgur.com/DbcoQli.png width=400>

Try experimenting with the built in colour scales in Plotly. You can find a full list [here](https://plotly.com/python/builtin-colorscales/).

* Find a way to set the colour scale using the color_continuous_scale parameter.
* Find a way to make the color axis disappear by using coloraxis_showscale.

In [61]:
bar = px.bar(x = num_genres.index[:15], # index = category name
             y = num_genres.values[:15],
             title = 'Top Genres',
             hover_name=num_genres.index[:15],
             color=num_genres.values[:15],
             color_continuous_scale = 'Agsunset')

bar.show()

In [62]:
bar.update_layout(xaxis_title='Genre',
yaxis_title='Number of Apps',
coloraxis_showscale=False)

bar.show()

# Grouped Bar Charts: Free vs. Paid Apps per Category

In [63]:
df_apps_clean.Type.value_counts()

,count
Type,
Free,7595
Paid,589


In [64]:
df_free_vs_paid = df_apps_clean.groupby(["Category", "Type"], as_index=False).agg({'App': pd.Series.count})
df_free_vs_paid.head()

,Category,Type,App
0,ART_AND_DESIGN,Free,58
1,ART_AND_DESIGN,Paid,3
2,AUTO_AND_VEHICLES,Free,72
3,AUTO_AND_VEHICLES,Paid,1
4,BEAUTY,Free,42


We see that the majority of apps are free on the Google Play Store. But perhaps some categories have more paid apps than others. Let’s investigate. We can group our data first by Category and then by Type. Then we can add up the number of apps per each type. Using as_index=False we push all the data into columns rather than end up with our Categories as the index.

**Challenge**: Use the plotly express bar [chart examples](https://plotly.com/python/bar-charts/#bar-chart-with-sorted-or-ordered-categories) and the [.bar() API reference](https://plotly.com/python-api-reference/generated/plotly.express.bar.html#plotly.express.bar) to create this bar chart:

<img src=https://imgur.com/LE0XCxA.png>

You'll want to use the `df_free_vs_paid` DataFrame that you created above that has the total number of free and paid apps per category.

See if you can figure out how to get the look above by changing the `categoryorder` to 'total descending' as outlined in the documentation here [here](https://plotly.com/python/categorical-axes/#automatically-sorting-categories-by-name-or-total-value).

In [65]:
g_bar = px.bar(df_free_vs_paid,
               x=df_free_vs_paid.Category,
               y=df_free_vs_paid.App,
               title='Free vs Paid Apps by Category',
               color='Type',
               barmode='group')


In [66]:
g_bar.update_layout(xaxis_title='Category',
                    yaxis_title='Number of Apps',
                    xaxis={'categoryorder':'total descending'},
                    yaxis=dict(type='log'),
                    xaxis_tickangle=90)

g_bar.show()

**Outcome**: What we see is that while there are very few paid apps on the Google Play Store, some categories have relatively more paid apps than others, including Personalization, Medical and Weather. So, depending on the category you are targeting, it might make sense to release a paid-for app.

# Plotly Box Plots: Lost Downloads for Paid Apps

**Challenge**: Create a box plot that shows the number of Installs for free versus paid apps. How does the median number of installations compare? Is the difference large or small?

Use the [Box Plots Guide](https://plotly.com/python/box-plots/) and the [.box API reference](https://plotly.com/python-api-reference/generated/plotly.express.box.html) to create the following chart.

<img src=https://imgur.com/uVsECT3.png>


In [116]:
fig = px.box(df_apps_clean, x= "Type", y="Installs", points="all", notched = True, color="Type")
fig.update_yaxes(
    tickmode = 'array',
    tickvals = [1, 100, 10000, 1000000, 100000000],
)

fig.update_layout(yaxis=dict(type='log'),title = 'How Many Downloads are Paid Giving Up?',)
fig.show()


**Outcome**: From the hover text in the chart, we see that the median number of downloads for free apps is 500,000, while the median number of downloads for paid apps is around 5,000! This is massively lower.But does this mean we should give up on selling a paid app? Let’s see how much revenue we would estimate per category.

# Plotly Box Plots: Revenue by App Category

**Challenge**: See if you can generate the chart below:

<img src=https://imgur.com/v4CiNqX.png>

Looking at the hover text, how much does the median app earn in the Tools category? If developing an Android app costs $30,000 or thereabouts, does the average photography app recoup its development costs?

Hint: I've used 'min ascending' to sort the categories.

In [121]:
df_paid_apps = df_apps_clean[df_apps_clean['Type'] == 'Paid']

fig = px.box(df_paid_apps, x= "Category", y="Revenue_Estimate",title='How Much Can Paid Apps Earn?' )
fig.update_yaxes(
    tickmode = 'array',
    tickvals = [1, 100, 10000, 1000000, 100000000],
)

fig.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Ballpark Revenue',
                  xaxis={'categoryorder':'min ascending'},
                  yaxis=dict(type='log'))

fig.show()


**Outcome**: If an Android app costs 30,000 dollar to develop, then the average app in very few categories would cover that development cost. The median paid photography app earned about $20,000. Many more app’s revenues were even lower - meaning they would need other sources of revenue like advertising or in-app purchases to make up for their development costs. However, certain app categories seem to contain a large number of outliers that have much higher (estimated) revenue - for example in Medical, Personalisation, Tools, Game, and Family.

# How Much Can You Charge? Examine Paid App Pricing Strategies by Category

**Challenge**: What is the median price price for a paid app? Then compare pricing by category by creating another box plot. But this time examine the prices (instead of the revenue estimates) of the paid apps. I recommend using `{categoryorder':'max descending'}` to sort the categories.

In [125]:
box = px.box(df_paid_apps,
             x='Category',
             y="Price",
             title='Price per Category')

In [129]:
box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Price',
                  xaxis={'categoryorder':'max descending'},
                  yaxis=dict(type='log'))

box.show()

**Outcome**: The median price for an Android app is 2.99 dollar. However, some categories have higher median prices than others. This time we see that Medical apps have the most expensive apps as well as a median price of 5.49 dollar. In contrast, Personalisation apps are quite cheap on average at 1.49 dollar . Other categories which higher median prices are Business (4.99 dollar) and Dating (6.99 dollar). It seems like customers who shop in these categories are not so concerned about paying a bit extra for their apps.